thrust/thrust/detail/contiguous_storage.inl

```diff
@@ -338,13 +343,18 @@ _CCCL_HOST_DEVICE void contiguous_storage<T, Alloc>::swap_allocators(true_type,
 template <typename T, typename Alloc>
 _CCCL_HOST_DEVICE void contiguous_storage<T, Alloc>::swap_allocators(false_type, Alloc& other)
 {
+  // FIXME(bgruber): it is really concerning, that swapping an allocator can throw. swap() should be noexcept in
+  // general.
   NV_IF_TARGET(NV_IS_DEVICE,
                (
                  // allocators must be equal when swapping containers with allocators that propagate on swap
                  assert(!is_allocator_not_equal(other));),
                (if (is_allocator_not_equal(other)) { throw allocator_mismatch_on_swap(); }));
```

Comment:
I have no ideas about the depths of allocators, but throwing an exception when swapping allocators strikes me as odd.

The [reference for `std::vector<T>::swap`](https://en.cppreference.com/w/cpp/container/vector/swap) says: 

> [...] if `get_allocator() != other.get_allocator()`, the behavior is undefined

I would rather prefer to have a hard fault here, so we can mark all swap functions of thrust vectors as `noexcept`.

Reply:
I mean swapping is move construction + move assignment so it can always potentially throw.

I would say it should not throw from our side, but especially allocators are potentially throwing a ton.

As an example think about a block allocator that reserves a huge slab of memory on construction and hands out slices of it. That could easily fail with an OOM

Reply:
> I mean swapping is move construction + move assignment so it can always potentially throw.

For a generic implementation, yes. In the long term, relocation could be used to avoid the moved from state, which could simplify things.

> As an example think about a block allocator that reserves a huge slab of memory on construction and hands out slices of it. That could easily fail with an OOM

How can swapping two block allocators using memory slabs fail with an OOM? I would still just swap pointers and some bookkeeping?

Reply:
I made some cleanup to the handling of allocators as a follow-up PR: #3050